### PR TITLE
ddl: release memory when adding index on partitions finish

### LINKED ISSUE #39189: Add index with ingest method does not work well on partition tables
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Create a partition table with a lot of partitions(>16).
2. Insert random data.
3. set @@global.tidb_ddl_enable_fast_reorg = 1;
4. Create an index on the partition table.
5. Check the history DDL by `ADMIN SHOW DDL JOBS`.

### 2. What did you expect to see? (Required)

It should use the ingest method.

In the field "job_type", it should be `add index /* ingest */`.

### 3. What did you see instead (Required)

`add index /* txn-merge */`

### 4. What is your TiDB version? (Required)

d00c588c04

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #39189

Problem Summary:

The tables with a lot of partitions cannot use the ingest method to create index. This is because 
1. `newAddIndexWorker()` increases the writer count and never decrease it until the engine is unregistered.
2. The backend and engine is reused for different partitions.
3. An error is reported when writer count becomes too large, and TiDB tries to fallback to the "txn-merge" method.

### What is changed and how it works?

Reset the writer counter and memory usage when a partition is done.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by xiongjiwei — APPROVED** [review]

File: ddl/ingest/engine_mgr.go [hunk 1]
```diff
@@ -99,6 +101,18 @@ func (m *engineManager) Unregister(jobID, indexID int64) {
...
 
+// ResetWorkers reset the writer count of the engineInfo because
+// the goroutines of backfill workers have been terminated.
+func (m *engineManager) ResetWorkers(jobID, indexID int64) {
+	ei, exist := m.Load(indexID)
+	if !exist {
+		return
+	}
+	m.MemRoot.Release(StructSizeWriterCtx * int64(ei.writerCount))
+	m.MemRoot.ReleaseWithTag(encodeEngineTag(jobID, indexID))
```
Comment (Benjamin2037) [inline ddl/ingest/engine_mgr.go:112]: Here release all mem use by engine and writer， but when start to handle next partition when regist engine and writer again, the engine releated  mem will not be count due to the exist of engine.

Comment (tangenta) [reply ddl/ingest/engine_mgr.go:112]: Nope, here we don't release engine struct memory(specifically, `StructSizeEngineInfo`). When all the backfill worker exit, `engineCacheSize` should be clear from the memory root.

**Review by Benjamin2037 — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**